packages/patterns/src/store/actions.js

```diff
@@ -47,18 +41,43 @@ export const createPattern =
 			.dispatch( coreStore )
 			.saveEntityRecord( 'postType', 'wp_block', reusableBlock );
 
-		if ( syncType === 'unsynced' || ! clientIds ) {
-			return updatedRecord;
+		return updatedRecord;
+	};
+
+/**
+ * Create a pattern from a JSON file.
+ * @param {File} file The JSON file instance of the pattern.
+ */
+export const createPatternFromFile =
+	( file ) =>
+	async ( { dispatch } ) => {
+		const fileContent = await file.text();
+		/** @type {import('./types').PatternJSON} */
+		let parsedContent;
+		try {
+			parsedContent = JSON.parse( fileContent );
+		} catch ( e ) {
+			throw new Error( 'Invalid JSON file' );
+		}
+		if (
+			parsedContent.__file !== 'wp_block' ||
+			! parsedContent.title ||
+			! parsedContent.content ||
+			typeof parsedContent.title !== 'string' ||
+			typeof parsedContent.content !== 'string' ||
+			( parsedContent.syncStatus &&
+				typeof parsedContent.syncStatus !== 'string' )
+		) {
+			throw new Error( 'Invalid Pattern JSON file' );
 		}
 
-		const newBlock = createBlock( 'core/block', {
-			ref: updatedRecord.id,
-		} );
-		registry
-			.dispatch( blockEditorStore )
-			.replaceBlocks( clientIds, newBlock );
-		dispatch.setEditingPattern( newBlock.clientId, true );
-		return updatedRecord;
+		const pattern = await dispatch.__experimentalCreatePattern(
```

Comment:
```suggestion
		const pattern = await dispatch.createPattern(
```

This needs updating after the rebase as well.

<img width="380" alt="Screenshot 2023-09-20 at 3 12 58 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/4123b534-15e4-4fe8-83c7-b9a458c93949">